src/uint/boxed/shl.rs

```diff
@@ -10,27 +10,37 @@ impl BoxedUint {
     /// Returns a zero and a truthy `Choice` if `shift >= self.bits_precision()`,
     /// or the result and a falsy `Choice` otherwise.
     pub fn shl(&self, shift: u32) -> (Self, Choice) {
+        let mut result = self.clone();
+        let overflow = result.shl_assign_with_carry(shift);
+        (result, overflow)
+    }
+
+    /// Computes `self <<= shift`.
+    ///
+    /// Returns a zero and a truthy `Choice` if `shift >= self.bits_precision()`,
+    /// or the result and a falsy `Choice` otherwise.
+    fn shl_assign_with_carry(&mut self, shift: u32) -> Choice {
```

Comment:
I don't think that "with_carry" is appropriate here, it doesn't return the bits that were shifted out of bounds. The returned `Choice` indicates the `shift` overflow.

Reply:
`with_overflow` then? Or other suggestions?

Reply:
Why not just `shl_assign`?

Reply:
@fjarri similar concern to [this comment](https://github.com/RustCrypto/crypto-bigint/pull/423#discussion_r1426092711) - I think it's confusing to have both an inherent method and a trait method with the same name but different type signatures. Where they do overlap I think they should have the same shape

Reply:
Then probably `overflowing_shl()` would be best. Although to match the method with the same name of the primitive types we would need to do `shift = shift % bits_precision`

Reply:
It already does: [L17 below](https://github.com/RustCrypto/crypto-bigint/pull/423/files#diff-f78b90032333fb7f636676d68f9ad9dc0a504a3d44c31561540d0c1dbd054443L17)

Reply:
Also good call on `overflowing_shl*`, renamed in [fee6c3f](https://github.com/RustCrypto/crypto-bigint/pull/423/commits/fee6c3f4a041202dc956ec198f16d9a26a43cbe7).

I could also rename `shl` above to `overflowing_shl` to match (since `shl` would still be available via the `Shl` trait).

Reply:
(I'd probably prefer to do that in a followup PR though to keep this one focused)